common/src/main/java/org/opensearch/ml/common/indexInsight/IndexInsightTask.java

```diff
@@ -0,0 +1,233 @@
+                break;
+            case COMPLETED:
+                // Check if needs update
+                if (lastUpdateTime != null && (currentTime - lastUpdateTime) > INDEX_INSIGHT_UPDATE_INTERVAL) {
+                    beginGeneration(storageIndex, tenantId, listener);
+                } else {
+                    // Return existing result
+                    IndexInsight insight = IndexInsight
+                        .builder()
+                        .index((String) source.get(IndexInsight.INDEX_NAME_FIELD))
+                        .taskType(MLIndexInsightType.valueOf((String) source.get(IndexInsight.TASK_TYPE_FIELD)))
+                        .content((String) source.get(IndexInsight.CONTENT_FIELD))
+                        .status(IndexInsightTaskStatus.COMPLETED)
+                        .lastUpdatedTime(Instant.ofEpochMilli(lastUpdateTime))
+                        .build();
+                    listener.onResponse(insight);
+                }
+                break;
+            case FAILED:
+                // Retry failed task
+                beginGeneration(storageIndex, tenantId, listener);
+                break;
+        }
+    }
+
+    /**
+     * Begin the index insight generation process by updating task status to GENERATING and executing the task with prerequisites.
+     */
+    default void beginGeneration(String storageIndex, String tenantId, ActionListener<IndexInsight> listener) {
+        String docId = generateDocId();
+        Map<String, Object> docMap = new HashMap<>();
+        docMap.put(IndexInsight.INDEX_NAME_FIELD, getSourceIndex());
+        docMap.put(IndexInsight.TASK_TYPE_FIELD, getTaskType().toString());
+        docMap.put(IndexInsight.STATUS_FIELD, IndexInsightTaskStatus.GENERATING.toString());
+        docMap.put(IndexInsight.LAST_UPDATE_FIELD, Instant.now().toEpochMilli());
+
+        UpdateRequest updateRequest = new UpdateRequest(storageIndex, docId).doc(docMap, MediaTypeRegistry.JSON).docAsUpsert(true);
+
+        getClient().update(updateRequest, ActionListener.wrap(r -> { runWithPrerequisites(storageIndex, tenantId, listener); }, e -> {
+            saveFailedStatus(storageIndex);
+            listener.onFailure(e);
+        }));
+    }
+
+    default void runWithPrerequisites(String storageIndex, String tenantId, ActionListener<IndexInsight> listener) {
+        List<MLIndexInsightType> prerequisites = getPrerequisites();
+        AtomicInteger completedCount = new AtomicInteger(0);
+        if (prerequisites.isEmpty()) {
+            runTask(storageIndex, tenantId, listener);
+            return;
+        }
+
+        // Run all prerequisites
+        for (MLIndexInsightType prerequisite : prerequisites) {
+            IndexInsightTask prerequisiteTask = createPrerequisiteTask(prerequisite);
+            prerequisiteTask.execute(storageIndex, tenantId, ActionListener.wrap(prereqInsight -> {
+                if (completedCount.incrementAndGet() == prerequisites.size()) {
```

Comment:
Is it possible that pre-requisite tasks have dependency? e.g. A depends B depends C and we need to run them in sequential order?

Reply:
Not Currently. And if A depends B depends C, we could define the A should only contain B in its prerequisite.